### PR TITLE
AP-3150: Remove multiple Guava library versions

### PR DESCRIPTION
**Pull request by raboczi**
This PR must be reviewed and merged together with its ApromoreEE counterpart https://github.com/apromore/ApromoreEE/pull/211.  It's the second PR for AP-3150 which upgrades various library versions as suggested by GitHub's Dependabot tool (https://github.com/apromore/ApromoreCore/security/dependabot).

- Centralize various library versions, pushing the version definitions upwards towards ApromoreCore/pom.xml and allowing them to be inherited consistently by the components below.  In particular, Guava 27.1-jre is now the only version of Guava present.
- Several unused library binaries (including Guava 16) were removed from the OpenXES source tree.
- Removed liquibase-osgi from core-bom; liquibase-core is already an OSGi bundle.
- Bump the `version.number` in site.properties from 7.19 to 7.20-SNAPSHOT.

### CODE REVIEW

**Comment by brucenguyen1** [comment]
This PR is a bit mix of different changes for different libraries: some library versions are not really changed but only version reference moved from plugin pom to central pom, some libraries (like Guava) does change the version which likely has a big impact.
Can we divide this change into smaller separate ones: one for version management and one for library change, and apply them one at a time, not all at once? Combining different types of change in one PR making it complex to go forward.

**Comment by brucenguyen1** [comment]
Regarding Guava change from different versions to v.27, I haven't seen any explicit review done in this PR regarding how it is being used by different modules in Apromore as a rationale that it is safe and no impacts when making this switch to Guava version 27 (at least we need a list of bullet points on what packages in Guava are used by what plugin to assess any release change impact). Due to the lack of this review, the question is how can this change be reviewed and made sure it doesn't cause issues in feature plugins, is it to test all the features in details?

File: site.properties [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 # Build tagging
 version.edition = @version.edition@
-version.number  = 7.19
+version.number  = 7.20
```
Comment (nolantellis) [inline site.properties]: I think it should be 7.20-development.
I am fine for the other changes

**Review by nolantellis — APPROVED** [review]

**Comment by raboczi** [comment]
> Regarding Guava change from different versions to v.27, I haven't seen any explicit review done in this PR regarding how it is being used by different modules in Apromore as a rationale that it is safe and no impacts when making this switch to Guava version 27 (at least we need a list of bullet points on what packages in Guava are used by what plugin to assess any release change impact). Due to the lack of this review, the question is how can this change be reviewed and made sure it doesn't cause issues in feature plugins, is it to test all the features in details?

To the best of my knowledge:
- Guava 14 used to be the default that Apromore code used, and was required because that's the exact version BoneCP wanted.  Now that Nolan has replaced BoneCP, we can successfully compile our own sources against Guava 27.  Assuming the semantics of Guava haven't been gratuitously changed such that the old library's methods do entirely different things from the new library's identically-named methods, I think we're entitled to believe this should work.  It's code we control.
- Similarly, we control the source of OpenXES which used to assume Guava 16.  Since we can recompile it successfully against Guava 27 and run our unit tests, that's pretty good evidence nothing has broken.
- ZK is a third party library with a transitive dependency on...some version of Guava, previously determined to be Guava 20.  This is the risky one, since it was almost certainly compiled against a different version of Guava than the one we're now using (27.1-jre).  This is our "run and pray" case, where we only truly know at runtime how back-compatible Guava is.
- Hadoop is a third party library which is dependent on Guava 27.1-jre; we have chosen to use the exact version it expects, so there should be no issue here.

[Cleaner] added:

- Apromore- Manager: Guava is used in unit tests only
  - EventLogServiceImplTest: com.google.common.collect.Sets
  - ProcessServiceImplUnitTest: com.google.common.io.CharStreams

- Apromore-Portal
  - AccessController: com.google.common.base.Strings
  - UserMenuController: com.google.common.base.Strings
  - MediaImpl: com.google.common.io.ByteStreams
  - ResourceServlet: com.google.common.io.ByteStreams
  - PortalPluginResourceServlet: com.google.common.io.ByteStreams
 
- CSV-Importer Logic: unit tests only
  - ParquetLogImporterCSVImplUnitTest: com.google.common.io.ByteStreams
  - TestUtilities: com.google.common.io.ByteStreams
 
- OpenXES:
  - XFactoryNaiveImpl: com.google.common.collect.Interner, com.google.common.collect.Interners

- SimilaritySearche-OSGI
  - GraphEditDistanceGreedy: com.google.common.collect.TreeMultimap, com.google.common.collect.TreeMultiset

- Log-Animation-Portal-Plugin: unit tests only.
  - TestDataSetup: com.google.common.collect.Lists

- Test-Tools
  - EqualsChecker: com.google.common.collect.Lists

File: site.properties (hunk 1 above)
Comment (raboczi) [reply site.properties]: I'll change it to 7.20-SNAPSHOT, since that's the GAV version coordinate.

**Comment by raboczi** [comment]
> Can we divide this change into smaller separate ones: one for version management and one for library change, and apply them one at a time, not all at once? Combining different types of change in one PR making it complex to go forward.

It's really all the same task: "make sure we only use one version of every library", with the changes to the pom.xml being the implementation detail of how we'd do this.  The removal of the unused SOAP/WSDL libraries is admittedly a separate issue I just tacked on while I was in the neighborhood.

**Comment by brucenguyen1** [comment]
> > Can we divide this change into smaller separate ones: one for version management and one for library change, and apply them one at a time, not all at once? Combining different types of change in one PR making it complex to go forward.
> 
> It's really all the same task: "make sure we only use one version of every library", with the changes to the pom.xml being the implementation detail of how we'd do this. The removal of the unused SOAP/WSDL libraries is admittedly a separate issue I just tacked on while I was in the neighborhood.

My recommendation is applying these changes in separate PRs because they are on different things and of different level of impact (you can always argue they are for the same goal and use that as a reason to combine them together, e.g. you can argue that all PRs in the repo are for "improve the system", this is however vague, you can always do things in steps and separately). If one PR has issue and has to revert back, the other PRs on different code/libs remain applicable.

**Comment by brucenguyen1** [comment]
> > Regarding Guava change from different versions to v.27, I haven't seen any explicit review done in this PR regarding how it is being used by different modules in Apromore as a rationale that it is safe and no impacts when making this switch to Guava version 27 (at least we need a list of bullet points on what packages in Guava are used by what plugin to assess any release change impact). Due to the lack of this review, the question is how can this change be reviewed and made sure it doesn't cause issues in feature plugins, is it to test all the features in details?
> 
> To the best of my knowledge:
> 
> * Guava 14 used to be the default that Apromore code used, and was required because that's the exact version BoneCP wanted.  Now that Nolan has replaced BoneCP, we can successfully compile our own sources against Guava 27.  Assuming the semantics of Guava haven't been gratuitously changed such that the old library's methods do entirely different things from the new library's identically-named methods, I think we're entitled to believe this should work.  It's code we control.
> * Similarly, we control the source of OpenXES which used to assume Guava 16.  Since we can recompile it successfully against Guava 27 and run our unit tests, that's pretty good evidence nothing has broken.
> * ZK is a third party library with a transitive dependency on...some version of Guava, previously determined to be Guava 20.  This is the risky one, since it was almost certainly compiled against a different version of Guava than the one we're now using (27.1-jre).  This is our "run and pray" case, where we only truly know at runtime how back-compatible Guava is.
> * Hadoop is a third party library which is dependent on Guava 27.1-jre; we have chosen to use the exact version it expects, so there should be no issue here.

I'll add to your above list more review on how Guava is being used in other plugins, but basically as it is "run and pray" with ZK, my recommendation is taking more time testing/using this PR before merging, and I would invite more reviewers from analysts (and it's not enough to assess it as compile success or not, the impact could be more subtle). This is not only a change of a library, it may affect features, shouldn't rush.

**Comment by raboczi** [comment]
> I'll add to your above list more review on how Guava is being used in other plugins, but basically as it is "run and pray" with ZK, my recommendation is taking more time testing/using this PR before merging, and I would invite more reviewers from analysts (and it's not enough to assess it as compile success or not, the impact could be more subtle). This is not only a change of a library, it may affect features, shouldn't rush.

I'll try to keep this branch up to date, in that case.  Dependabot is insisting on Guava 24 or later, so the original issue cannot be resolved without changing ZK from using Guava 20.

Here's the review of how Guava is being used in other plugins, at least on the Karaf assembly.  From the console running the development branch, this is what a query for all Guava package exporters and importers shows:
```
karaf@root()> package:exports -p com.google.common                                                                                                                                                                  
Package Name                               │ Version │ ID │ Bundle Name
───────────────────────────────────────────┼─────────┼────┼───────────────────────────────
com.google.common.annotations              │ 14.0.1  │ 29 │ com.google.guava
com.google.common.annotations              │ 16.0.1  │ 30 │ com.google.guava
com.google.common.annotations              │ 20.0.0  │ 31 │ com.google.guava
com.google.common.annotations              │ 27.1.0  │ 32 │ com.google.guava
com.google.common.base                     │ 14.0.1  │ 29 │ com.google.guava
com.google.common.base                     │ 16.0.1  │ 30 │ com.google.guava
com.google.common.base                     │ 20.0.0  │ 31 │ com.google.guava
com.google.common.base                     │ 27.1.0  │ 32 │ com.google.guava
com.google.common.cache                    │ 14.0.1  │ 29 │ com.google.guava
com.google.common.cache                    │ 16.0.1  │ 30 │ com.google.guava
com.google.common.cache                    │ 20.0.0  │ 31 │ com.google.guava
com.google.common.cache                    │ 27.1.0  │ 32 │ com.google.guava
com.google.common.collect                  │ 14.0.1  │ 29 │ com.google.guava
com.google.common.collect                  │ 16.0.1  │ 30 │ com.google.guava
com.google.common.collect                  │ 20.0.0  │ 31 │ com.google.guava
com.google.common.collect                  │ 27.1.0  │ 32 │ com.google.guava
com.google.common.escape                   │ 16.0.1  │ 30 │ com.google.guava
com.google.common.escape                   │ 20.0.0  │ 31 │ com.google.guava
com.google.common.escape                   │ 27.1.0  │ 32 │ com.google.guava
com.google.common.eventbus                 │ 14.0.1  │ 29 │ com.google.guava
com.google.common.eventbus                 │ 16.0.1  │ 30 │ com.google.guava
com.google.common.eventbus                 │ 20.0.0  │ 31 │ com.google.guava
com.google.common.eventbus                 │ 27.1.0  │ 32 │ com.google.guava
com.google.common.graph                    │ 20.0.0  │ 31 │ com.google.guava
com.google.common.graph                    │ 27.1.0  │ 32 │ com.google.guava
com.google.common.hash                     │ 14.0.1  │ 29 │ com.google.guava
com.google.common.hash                     │ 16.0.1  │ 30 │ com.google.guava
com.google.common.hash                     │ 20.0.0  │ 31 │ com.google.guava
com.google.common.hash                     │ 27.1.0  │ 32 │ com.google.guava
com.google.common.html                     │ 16.0.1  │ 30 │ com.google.guava
com.google.common.html                     │ 20.0.0  │ 31 │ com.google.guava
com.google.common.html                     │ 27.1.0  │ 32 │ com.google.guava
com.google.common.io                       │ 14.0.1  │ 29 │ com.google.guava
com.google.common.io                       │ 16.0.1  │ 30 │ com.google.guava
com.google.common.io                       │ 20.0.0  │ 31 │ com.google.guava
com.google.common.io                       │ 27.1.0  │ 32 │ com.google.guava
com.google.common.math                     │ 14.0.1  │ 29 │ com.google.guava
com.google.common.math                     │ 16.0.1  │ 30 │ com.google.guava
com.google.common.math                     │ 20.0.0  │ 31 │ com.google.guava
com.google.common.math                     │ 27.1.0  │ 32 │ com.google.guava
com.google.common.net                      │ 14.0.1  │ 29 │ com.google.guava
com.google.common.net                      │ 16.0.1  │ 30 │ com.google.guava
com.google.common.net                      │ 20.0.0  │ 31 │ com.google.guava
com.google.common.net                      │ 27.1.0  │ 32 │ com.google.guava
com.google.common.primitives               │ 14.0.1  │ 29 │ com.google.guava
com.google.common.primitives               │ 16.0.1  │ 30 │ com.google.guava
com.google.common.primitives               │ 20.0.0  │ 31 │ com.google.guava
com.google.common.primitives               │ 27.1.0  │ 32 │ com.google.guava
com.google.common.reflect                  │ 14.0.1  │ 29 │ com.google.guava
com.google.common.reflect                  │ 16.0.1  │ 30 │ com.google.guava
com.google.common.reflect                  │ 20.0.0  │ 31 │ com.google.guava
com.google.common.reflect                  │ 27.1.0  │ 32 │ com.google.guava
com.google.common.util.concurrent.internal │ 1.0.1   │ 33 │ com.google.guava.failureaccess
com.google.common.util.concurrent          │ 14.0.1  │ 29 │ com.google.guava
com.google.common.util.concurrent          │ 16.0.1  │ 30 │ com.google.guava
com.google.common.util.concurrent          │ 20.0.0  │ 31 │ com.google.guava
com.google.common.util.concurrent          │ 27.1.0  │ 32 │ com.google.guava
com.google.common.xml                      │ 16.0.1  │ 30 │ com.google.guava
com.google.common.xml                      │ 20.0.0  │ 31 │ com.google.guava
com.google.common.xml                      │ 27.1.0  │ 32 │ com.google.guava

karaf@root()> package:imports -p com.google.common                                                                                                                                                                  
Package                                    │ Version         │ Optional │ ID  │ Bundle Name
───────────────────────────────────────────┼─────────────────┼──────────┼─────┼───────────────────────────────────
com.google.common.base                     │ [14.0.0,15.0.0) │          │ 209 │ org.apromore.portal
com.google.common.base                     │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.cache                    │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.collect                  │ [14.0.0,15.0.0) │          │ 187 │ org.apromore.log-osgi
com.google.common.collect                  │ [14.0.0,15.0.0) │          │ 218 │ org.apromore.similaritysearch-osgi
com.google.common.collect                  │ [20.0.0,21.0.0) │          │ 226 │ org.apromore.zk-osgi-ee
com.google.common.collect                  │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.io                       │ [14.0.0,15.0.0) │ resolved │ 209 │ org.apromore.portal
com.google.common.io                       │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.net                      │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.primitives               │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
com.google.common.util.concurrent.internal │ [1.0.0,2.0.0)   │          │ 32  │ com.google.guava
com.google.common.util.concurrent          │ [27.0.0,28.0.0) │          │ 193 │ org.apromore.parquet-osgi
```
So what this is telling us is that:
- The various Guava packages are each available in four different versions (except for com.google.commons.xml which v14 is too early to include) from the Guava bundles numbered 29 to 32.
- Guava v20 is only used by ZK, Guava v27 is only used by Hadoop, Guava v14 is used by the remaining clients (the Portal, the Log library [OpenXES], Similarity Search), and Guava v16 isn't actually used at all.
- Guava v27 has one internal dependency provided by the bundle com.google.guava.failureaccess; we can safely ignore that.

And this is what it looks like under this PR's branch (together with its EE counterpart):
```
karaf@root()> package:exports -p com.google.common                                                                                                                                     
Package Name                               │ Version │ ID │ Bundle Name
───────────────────────────────────────────┼─────────┼────┼───────────────────────────────
com.google.common.annotations              │ 27.1.0  │ 29 │ com.google.guava
com.google.common.base                     │ 27.1.0  │ 29 │ com.google.guava
com.google.common.cache                    │ 27.1.0  │ 29 │ com.google.guava
com.google.common.collect                  │ 27.1.0  │ 29 │ com.google.guava
com.google.common.escape                   │ 27.1.0  │ 29 │ com.google.guava
com.google.common.eventbus                 │ 27.1.0  │ 29 │ com.google.guava
com.google.common.graph                    │ 27.1.0  │ 29 │ com.google.guava
com.google.common.hash                     │ 27.1.0  │ 29 │ com.google.guava
com.google.common.html                     │ 27.1.0  │ 29 │ com.google.guava
com.google.common.io                       │ 27.1.0  │ 29 │ com.google.guava
com.google.common.math                     │ 27.1.0  │ 29 │ com.google.guava
com.google.common.net                      │ 27.1.0  │ 29 │ com.google.guava
com.google.common.primitives               │ 27.1.0  │ 29 │ com.google.guava
com.google.common.reflect                  │ 27.1.0  │ 29 │ com.google.guava
com.google.common.util.concurrent.internal │ 1.0.1   │ 30 │ com.google.guava.failureaccess
com.google.common.util.concurrent          │ 27.1.0  │ 29 │ com.google.guava
com.google.common.xml                      │ 27.1.0  │ 29 │ com.google.guava

karaf@root()> package:imports -p com.google.common                                                                                                                                                                  
Package                                    │ Version         │ Optional │ ID  │ Bundle Name
───────────────────────────────────────────┼─────────────────┼──────────┼─────┼───────────────────────────────────
com.google.common.base                     │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.base                     │ [27.1.0,28.0.0) │          │ 206 │ org.apromore.portal
com.google.common.cache                    │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.collect                  │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.collect                  │ [27.1.0,28.0.0) │          │ 184 │ org.apromore.log-osgi
com.google.common.collect                  │ [27.1.0,28.0.0) │          │ 215 │ org.apromore.similaritysearch-osgi
com.google.common.collect                  │ [27.1.0,28.0.0) │          │ 223 │ org.apromore.zk-osgi-ee
com.google.common.io                       │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.io                       │ [27.1.0,28.0.0) │ resolved │ 206 │ org.apromore.portal
com.google.common.net                      │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.primitives               │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
com.google.common.util.concurrent.internal │ [1.0.0,2.0.0)   │          │ 29  │ com.google.guava
com.google.common.util.concurrent          │ [27.0.0,28.0.0) │          │ 190 │ org.apromore.parquet-osgi
```
All the same clients, but only bundle 29 (Guava 27.1-jre) is exporting the packages this time.

**Comment by brucenguyen1** [comment]
It's not enough information from looking at the package names only, such as com.google.common.collect or io.

From my source code review above, the specific Guava classes used by Apromore plugins are (under com.google.common package)
- base.Strings
- io.CharStreams
- io.ByteStreams
- collect.Interner
- collect.Interners
- collect.Sets
- collect.TreeMultimap
- collect.TreeMultiset
- collect.Lists

These are String utilities, some io and some collection classes. The main question is whether these class interfaces changed when moving from old versions to v27 (i.e. the public class members actually used by Apromore). Probably they rarely change, but it's not clear until we can read version release notes and/or testing because we don't want to build and run a system based on luck.

I don't know about the above information regarding ZK and Parquet, but it's possible to review their source code to know the extent of Guava classes they use and so whether it is safe to move from old versions of Guava to v27.

**Comment by raboczi** [comment]
> I don't know about the above information regarding ZK and Parquet, but it's possible to review their source code to know the extent of Guava classes they use and so whether it is safe to move from old versions of Guava to v27.

In the case where we've compiled the bundle from source (that's OpenXES, Portal, Similarity Search) we're guaranteed to be compatible.  Parquet is guaranteed because we're giving it the exact version it advertises.  So that leaves ZK for be confirmed.  If we were to compile ZK from source successfully with Guava 27.1, would that persuade you that ZK contains no incompatible API calls?

**Comment by brucenguyen1** [comment]
> > I don't know about the above information regarding ZK and Parquet, but it's possible to review their source code to know the extent of Guava classes they use and so whether it is safe to move from old versions of Guava to v27.
> 
> In the case where we've compiled the bundle from source (that's OpenXES, Portal, Similarity Search) we're guaranteed to be compatible. Parquet is guaranteed because we're giving it the exact version it advertises. So that leaves ZK for be confirmed. If we were to compile ZK from source successfully with Guava 27.1, would that persuade you that ZK contains no incompatible API calls?

Hi Simon, as said above, compile success or not is not enough. There could be subtle changes regarding the semantic, performance and extent of each interface. This is software code and it can be changed in any way. After all of this discussion, it sounds like I'm the barrier here. If so, just go ahead. You can merge them at any time.

**Pull request merged**